Add config flow to Switchbot

RenierM26 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->



## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
-Add config_flow to switchbot integration. 
The format also allows for easy adding of additional platforms in future pull requests.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [X] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/19371

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

RenierM26 committed:
Add config_flow and cover platform.

RenierM26 committed:
Add test for config_flow

RenierM26 committed:
pylint import exclude added in cover.py, bump api version to latest.

frenck commented on homeassistant/components/switchbot/cover.py:
```diff
@@ -0,0 +1,149 @@
+"""Support for SwitchBot curtains."""
```
This is adding a new platform, which is out of scope for this PR.

Please make a separate PR for the configuration flow and one for adding the cover.

RenierM26 replied:
Hi @frenck ,

I have removed the cover platform as requested.

RenierM26 committed:
Remove cover platform, add bluetooth device connection test to config_flow.

RenierM26 committed:
Fix tests, improve config_flow bluetooth connection test and exceptions.

RenierM26 committed:
Trying to fix weird import error.

RenierM26 committed:
Requirements file has # next to switchbot imports, causing pylint import error. Added disable import error to tests as well.

RenierM26 committed:
Remove bluepy from config_flow, switchbot api has private methods to test connectivity to device.

RenierM26 committed:
Add test for user input without password.

RenierM26 committed:
Update requirements_test_all.txt

For some reason the gen requirements script hashes out any switch-bot requirements. Importing the API is a requirement for config_flow tests to pass.

RenierM26 commented on requirements_test_all.txt:
```diff
@@ -23,6 +23,9 @@ PyQRCode==1.2.1
 # homeassistant.components.rmvtransport
 PyRMVtransport==0.3.2
 
+# homeassistant.components.switchbot
+PySwitchbot==0.10.0
+
```
The generate requirements script hashes out "PySwitchbot==0.10.0". The config_flow tests require this import.

Config flow uses the api to test connection to device. I also test those lines in the config_flow test by mocking the pyswitchbot api.

frenck replied:
pySwitchbot requires bluetooth/bluepy, and is therefore cannot be used for that. It will be automatically commented out for that reason.

RenierM26 replied:
@ frenck,

Found a workaround by using pygatt to test connection to device. (same library as hass Bluetooth LE component). 

Thanks.

RenierM26 committed:
Add mock class and method to tests to replace import of pyswitchbot

RenierM26 committed:
Cleanup test fixture

RenierM26 committed:
Remove device connection test from config flow as there's no way to test bluetooth api.

RenierM26 committed:
Use pygatt to test btle device connectivity, updates tests.

RenierM26 committed:
Removed duplicate abort if unique id configured.

RenierM26 committed:
Fix entry setup.

RenierM26 committed:
Cleanup __init__.py

RenierM26 committed:
Add switch device_info

RenierM26 commented on homeassistant/components/switchbot/config_flow.py:
```diff
@@ -0,0 +1,92 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+def _btle_connect(mac):
+    """Try to connect to switchbot device."""
+    adapter = pygatt.GATTToolBackend()
+
+    try:
+        adapter.start()
+        device = adapter.connect(mac, address_type=pygatt.BLEAddressType.random)
+    finally:
+        adapter.stop()
+
+    return device
+
```
Found a potential problem with pygatt. Let me investigate.

RenierM26 replied:
Found and fixed the issue. Seems to be related to adapter.start() that tries to run a command that doesn't exist. 

This fixes the problem : adapter.start(reset_on_start=False)

RenierM26 committed:
Fix pygatt command not found error on adapter.start

RenierM26 committed:
Test mock import module, bump api version for "discovery" used to test user imput.

RenierM26 committed:
fix .coveragerc

RenierM26 committed:
.coveragerc

RenierM26 committed:
Added config_flow input validation, basic typing, cleanup all references to other switchbot device types. A few reliablity improvements (for cheap bt adapters).

RenierM26 commented:
Hi @Danielhiversen,

I'm done with my hass pull request adding config_flow and update mechanics to the switchbot integration. Mind having a look if you get a chance?

Thank you!

Danielhiversen commented on homeassistant/components/switchbot/switch.py:
```diff
@@ -1,18 +1,33 @@
-DEFAULT_NAME = "Switchbot"
+# Initialize the logger
+_LOGGER = logging.getLogger(__name__)
+CONNECT_LOCK = Lock()
```
Should this be the same lock as in coordinator.py?
Maybe include it in SwitchbotDataUpdateCoordinator?

RenierM26 replied:
I tried that and it then has no effect. The log gets spammed with update failures. 

If set outside of the class, it works. Reliable state updates without errors.

Danielhiversen commented on homeassistant/components/switchbot/switch.py:
```diff
@@ -23,59 +38,103 @@
 
 
-class SwitchBot(SwitchEntity, RestoreEntity):
+class SwitchBot(CoordinatorEntity, SwitchEntity):
```
Why do you remove `RestoreEntity`?

RenierM26 replied:
Readded RestoreEntity. I though this was only needed for the "assumed switch state". It doesn't seem to have any effect.

Danielhiversen commented on homeassistant/components/switchbot/switch.py:
```diff
@@ -23,59 +38,103 @@
         """Return true if unable to access real state of entity."""
-        return True
+        if not self.coordinator.data[self._idx]["data"]["switchMode"]:
+            return True
```
```suggestion
            return True
        return False
```

Danielhiversen replied:
```suggestion
            return True
        return False
```

Danielhiversen commented on homeassistant/components/switchbot/coordinator.py:
```diff
@@ -0,0 +1,61 @@
+        )
+
+        if self._switchbot_data:
+            return True
```
```suggestion
            return True
        return False
```

Danielhiversen commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,80 @@
+    switchbot.DEFAULT_RETRY_TIMEOUT = entry.options[CONF_RETRY_TIMEOUT]
+
+    # Store api in coordinator.
+    coordinator = SwitchbotDataUpdateCoordinator(
+        hass,
+        update_interval=entry.options[CONF_TIME_BETWEEN_UPDATE_COMMAND],
+        api=switchbot,
+        retry_count=entry.options[CONF_RETRY_COUNT],
+        scan_timeout=entry.options[CONF_SCAN_TIMEOUT],
+    )
```
Why do you want to store the API in coordinator? Why not import it where needed?

RenierM26 replied:
1) To allow the user to change api defaults in config flow (Retry timeout L38 - L39)
2) For easy access in all platforms.

I could import just the GetSwitchbotDevices class and then import on the api on each platform. The coordinator is shared between all platforms so I think this is the most efficient way to do this. (Python just stores a memory reference in each platform I think.)

RenierM26 committed:
Update homeassistant/components/switchbot/switch.py

Co-authored-by: Daniel Hjelseth Høyer <mail@dahoiv.net>

RenierM26 committed:
Update homeassistant/components/switchbot/coordinator.py

Co-authored-by: Daniel Hjelseth Høyer <mail@dahoiv.net>

RenierM26 committed:
Import asyncio lock in coordinator, request refresh when operating switch, add restoreentity mixin to switch platform.

RenierM26 committed:
with asyncio.lock has no effect on platform services, removed.

RenierM26 committed:
Single coordinator for all entities, asyncio lock to prevent BTLE errors.

RenierM26 committed:
Rerun hassfest after rebase.

RenierM26 committed:
Removed sleep used for test.

Danielhiversen commented on .coveragerc:
```diff
@@ -1005,6 +1005,11 @@ omit =
+    homeassistant/components/switchbot/__init__.py
+    homeassistant/components/switchbot/const.py
+    homeassistant/components/switchbot/coordinator.py
+    homeassistant/components/switcher_kis/sensor.py
+    homeassistant/components/switcher_kis/switch.py
```
```suggestion
```

Danielhiversen commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,80 @@
+
+        # BTLE has issues with multiple connections,
+        # so we use a lock to ensure that only one API request is reaching it at a time:
+        api_lock = Lock()
```
Out of scope for this pr, but it would probably be useful to have this as a global lock in Home Assistant, so all integrations with bt could use the same lock

bdraco replied:
Maybe a central broker to obtain btle access.

Danielhiversen reviewed (approved):
LGTM 👍 

Great work

RenierM26 committed:
Update .coveragerc

Co-authored-by: Daniel Hjelseth Høyer <mail@dahoiv.net>

bdraco commented on homeassistant/components/switchbot/config_flow.py:
```diff
@@ -0,0 +1,172 @@
+
+        import_config[CONF_MAC] = import_config[CONF_MAC].replace("-", ":").lower()
+
+        await self.async_set_unique_id(import_config[CONF_MAC].replace(":", ""))
```
Can you use `format_mac` here?
https://developers.home-assistant.io/docs/entity_registry_index/#example-acceptable-sources-for-a-unique-id

RenierM26 replied:
format_mac won't work as I intentionally remove ":" in order to match with coordinator provided data.

RenierM26 committed:
Use format_mac for unique_id

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,80 @@
+async def async_unload_entry(hass, entry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
```
Need to cleanup  `hass.data[DOMAIN]` here

RenierM26 replied:
Added cleanup.

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,80 @@
+            raise ConfigEntryNotReady
+
+        # Creates reference to api via coordinator, this should result in single poll for all entities.
+        hass.data[DOMAIN] = {DATA_COORDINATOR: coordinator}
```
If they have more than one switchbot, and setup a second config entry this will shadow the first entry.

Usually we store it as `hass.data[DOMAIN][entry.entry_id]`

RenierM26 replied:
Storing coordinator as hass.data[DOMAIN][entry.entry_id]

bdraco reviewed (changes requested):
Please see comments above

RenierM26 committed:
__init__ change to store coordinator in hass.entity_id, complete typing.

RenierM26 commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,97 @@
+
+        hass.config_entries.async_update_entry(entry, options=options)
+
+    # Use same coordinator instance for all entities.
+    # Uses BTLE advertisement data, all Switchbot devices in range is stored here.
+    if hass.data.get(DOMAIN):
+        for item in hass.config_entries.async_entries(domain=DOMAIN):
+            if hass.data[DOMAIN].get(item.entry_id):
+                coordinator = hass.data[DOMAIN][item.entry_id].get(DATA_COORDINATOR)
+                break
```
1) Check if coordinator is already loaded by another entity setup.
2) If found, set coordinator to same object.

Data is obtained from BTLE advertisements provided from a scan. Data for all devices in proximity is returned. The API builds a dict with entities and their corresponding setting. (On/Off state and mode for the bots as example)

I'm going to add sensors for Battery etc. in future pull requests.

RenierM26 replied:
It's not possible to do more than one simultaneous scan per Bluetooth adapter and it's not needed. 

I've tested removing/adding/changing options of entities without any issues or lockups. Can even remove the first loaded entity and others still update just fine.

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,97 @@
+
+        hass.config_entries.async_update_entry(entry, options=options)
+
+    # Use same coordinator instance for all entities.
```
This looks like a lot of code to find the coordinator.

Would it be simpler to store the coordinator as and destroy it when the last config entry is unloaded?

hass.data[DOMAIN][DATA_COORDINATOR]

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,97 @@
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    undo_listener = entry.add_update_listener(_async_update_listener)
```
```suggestion
    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
```

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,97 @@
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        DATA_COORDINATOR: coordinator,
+        DATA_UNDO_UPDATE_LISTENER: undo_listener,
```
```suggestion
```

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,97 @@
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+    if unload_ok:
+        hass.data[DOMAIN][entry.entry_id][DATA_UNDO_UPDATE_LISTENER]()
```
```suggestion
```

bdraco commented on homeassistant/components/switchbot/const.py:
```diff
@@ -0,0 +1,23 @@
+
+# Data
+DATA_COORDINATOR = "coordinator"
+DATA_UNDO_UPDATE_LISTENER = "undo_update_listener"
```
```suggestion
```

bdraco commented on homeassistant/components/switchbot/switch.py:
```diff
@@ -85,9 +176,28 @@ def unique_id(self) -> str:
-        return {"last_run_success": self._last_run_success}
+        return {
+            "last_run_success": self._last_run_success,
+            "MAC": self._mac,
```
```suggestion
            "mac_address": self._mac,
```

bdraco commented on homeassistant/components/switchbot/switch.py:
```diff
@@ -85,9 +176,28 @@ def unique_id(self) -> str:
+    def device_info(self) -> DeviceInfo:
+        """Return the device_info of the device."""
+        return {
+            "identifiers": {(DOMAIN, self._mac.replace(":", ""))},
```
Use `connections` instead of `identifiers`


from homeassistant.helpers import device_registry as dr


```suggestion
            "connections": {(dr.CONNECTION_NETWORK_MAC, self._mac)},
```

bdraco commented on homeassistant/components/switchbot/coordinator.py:
```diff
@@ -0,0 +1,59 @@
+from datetime import timedelta
+import logging
+
+import switchbot  # pylint: disable=import-error
```
```suggestion
import switchbot
```

This should't be needed

RenierM26 replied:
The library is "force commented" from the requirements file. According to @frenck this is done due to issues with hassio commit tests workflow.

bdraco replied:
Do you know what the issue is? Can we fix it in the upstream library to avoid this problem?

RenierM26 replied:
I think it's related to bluepy. I see all of the integrations making use of bluepy are commented out.

# homeassistant.components.decora
# homeassistant.components.miflora
# bluepy==1.3.0

The PySwitchbot library makes use of bluepy.

bdraco commented on homeassistant/components/switchbot/config_flow.py:
```diff
@@ -0,0 +1,180 @@
+        # Validate bluetooth device mac.
+        # CONNECT_LOCK prevents btle adapter exceptions if there are multiple calls to this method.
+        try:
+            async with CONNECT_LOCK:
```
Store this in `hass.data[DOMAIN][BTLE_LOCK]` instead since it likely needs to share the same lock with the coordinator.

RenierM26 replied:
Done.

bdraco commented on homeassistant/components/switchbot/coordinator.py:
```diff
@@ -0,0 +1,59 @@
+            hass, _LOGGER, name=DOMAIN, update_interval=self.update_interval
+        )
+
+        self.api_lock = api_lock
```
Store this in `hass.data[DOMAIN][BTLE_LOCK]` since it likely needs to share the same lock as the config flow.

RenierM26 replied:
Done.

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,97 @@
+"""Support for Switchbot devices."""
+from asyncio import Lock
+
+import switchbot  # pylint: disable=import-error
```
```suggestion
import switchbot
```

This shouldn't be needed if we have the requirements set correctly in `manifest.json`

RenierM26 replied:
The library is "force commented" from the hassio requirements file. According to @frenck this is done due to issues with hassio commit tests workflow.

bdraco reviewed (changes requested):
Please see comments above

RenierM26 committed:
Update homeassistant/components/switchbot/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Update homeassistant/components/switchbot/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Update homeassistant/components/switchbot/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Update homeassistant/components/switchbot/const.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Update homeassistant/components/switchbot/switch.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Update homeassistant/components/switchbot/switch.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Update homeassistant/components/switchbot/switch.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Cleanup switch.py

RenierM26 committed:
Store coordinator if it doesn't exist, pass asyncio.lock from config_flow

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,97 @@
+
+    # Use same coordinator instance for all entities.
+    # Uses BTLE advertisement data, all Switchbot devices in range is stored here.
+    if not hass.data[DOMAIN].get(DATA_COORDINATOR):
```
```suggestion
    if DATA_COORDINATOR not in hass.data[DOMAIN]:
```

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,97 @@
+        # Check if asyncio.lock is stored in hass data.
+        # BTLE has issues with multiple connections,
+        # so we use a lock to ensure that only one API request is reaching it at a time:
+        if not hass.data[DOMAIN].get(BTLE_LOCK):
```
```suggestion
        if BTLE_LOCK not in hass.data[DOMAIN]:
```

bdraco commented on homeassistant/components/switchbot/config_flow.py:
```diff
@@ -0,0 +1,187 @@
+
+        # asyncio.lock prevents btle adapter exceptions if there are multiple calls to this method.
+        # store asyncio.lock in hass data if not present.
+        if not self.hass.data.get(DOMAIN):
```
```suggestion
        if DOMAIN not in self.hass.data:
```

bdraco commented on homeassistant/components/switchbot/config_flow.py:
```diff
@@ -0,0 +1,187 @@
+        # store asyncio.lock in hass data if not present.
+        if not self.hass.data.get(DOMAIN):
+            self.hass.data.setdefault(DOMAIN, {})
+            self.hass.data[DOMAIN][BTLE_LOCK] = Lock()
```
```suggestion
        if BTLE_LOCK not in self.hass.data[DOMAIN]:
            self.hass.data[DOMAIN][BTLE_LOCK] = Lock()
```

If there is ever a change in `async_setup_entry` that `await`s after creating the empty. dict in `hass.data[DOMAIN]`, it could return control to the event loop which would mean `hass.data[DOMAIN]` would exist but the BTLE_LOCK had not been created yet. This code would create the lock instead of `async_setup_entry`.  Right now there is a check in `async_setup_entry` to prevent creating it so it should be safe.  However without explicitly checking there is some risk we could clobber the lock here due to later refactoring.

bdraco commented on homeassistant/components/switchbot/config_flow.py:
```diff
@@ -0,0 +1,187 @@
+
+        connect_lock = self.hass.data[DOMAIN][BTLE_LOCK]
+
+        # Validate bluetooth device mac.
+        try:
+            async with connect_lock:
+                _btle_adv_data = await self.hass.async_add_executor_job(
+                    _btle_connect, data[CONF_MAC]
+                )
+
+        except NotConnectedError as err:
+            raise NotConnectedError(err) from err
```
```suggestion
        # Validate bluetooth device mac.
        async with connect_lock:
            _btle_adv_data = await self.hass.async_add_executor_job(
                _btle_connect, data[CONF_MAC]
            )

```

The try/except is re-raising the same named exception here.

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,97 @@
+        if not hass.data[DOMAIN].get(BTLE_LOCK):
+            hass.data[DOMAIN][BTLE_LOCK] = Lock()
+
+        switchbot.DEFAULT_RETRY_TIMEOUT = entry.options[CONF_RETRY_TIMEOUT]
```
If the user sets a different `CONF_RETRY_TIMEOUT` for second config entry, this it will clobber the value here.

RenierM26 replied:
Implemented a solution for this by storing the entry in hass.data[DOMAIN][CONF_RETRY_TIMEOUT] and updating options on  reload for all entities.

RenierM26 committed:
Update homeassistant/components/switchbot/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Update homeassistant/components/switchbot/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Update homeassistant/components/switchbot/config_flow.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Update homeassistant/components/switchbot/config_flow.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Update homeassistant/components/switchbot/config_flow.py

Co-authored-by: J. Nick Koston <nick@koston.org>

RenierM26 committed:
Reload all entities when retry timeout changes.

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,115 @@
+    if entry.options[CONF_RETRY_TIMEOUT] != hass.data[DOMAIN][CONF_RETRY_TIMEOUT]:
+        hass.data[DOMAIN][CONF_RETRY_TIMEOUT] = entry.options[CONF_RETRY_TIMEOUT]
+        hass.data[DOMAIN].pop(DATA_COORDINATOR)
+
```
At next startup, the other entries could still have a different timeout so you'll get whichever one is loaded first.

bdraco replied:
Maybe update the other config entries with the new value?

RenierM26 replied:
Added code to update config entries L92-L95.

bdraco commented:
Once the `CONF_RETRY_TIMEOUT` is addressed this should be good to approve 👍

RenierM26 committed:
Update entry options on RETRY TIMEOUT change.

RenierM26 committed:
Store all common options in hass, update common option when entity options change.

bdraco commented on homeassistant/components/switchbot/__init__.py:
```diff
@@ -1 +1,121 @@
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+    if unload_ok:
+        # Update entity options stored in hass.
```
I think it would be cleaner to do this when saving the options flow.  Update all other config entries at the same time except the one we are editing (let that one get updated from the flow result)

RenierM26 committed:
Rerun hassfest.

bdraco commented on homeassistant/components/switchbot/manifest.json:
```diff
@@ -2,7 +2,8 @@
-  "requirements": ["PySwitchbot==0.8.0"],
+  "requirements": ["PySwitchbot==0.11.0"],
+  "config_flow": true,
   "codeowners": ["@danielhiversen"],
```
Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

RenierM26 committed:
Update common entity options via options flow and cleanup init.

RenierM26 committed:
Update hass data on options update.

bdraco commented:
Test failure is unrelated. I restarted the CI.

bdraco commented:
In a future PR we should discover available switch bots and let the user choose instead of having them type in a MAC address.  In testing I typo-ed the mac address and it took me a few tried to figure out what went wrong.

bdraco commented:
Should be good to go once a docs PR is opened

RenierM26 commented:
Hi @bdraco,

Thank you for all your help with this pull request. Much appreciated!

I have opened a documentation pull request : https://github.com/home-assistant/home-assistant.io/pull/19371

RenierM26 committed:
Add test for entry options change.

bdraco reviewed (approved):
Thanks 👍
